File: src/csharp/pulse/cdm/scalars/SEScalarPressureTimePerVolume.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

public class PressureTimePerVolumeUnit
{
  private PressureTimePerVolumeUnit(string v) { Value = v; }
  protected string Value { get; }

  public new string ToString() { return Value; }

  public static PressureTimePerVolumeUnit cmH2O_s_Per_L = new PressureTimePerVolumeUnit("cmH2O s/L");
  public static PressureTimePerVolumeUnit cmH2O_s_Per_mL = new PressureTimePerVolumeUnit("cmH2O s/mL");
  public static PressureTimePerVolumeUnit mmHg_s_Per_mL = new PressureTimePerVolumeUnit("mmHg s/mL");
  public static PressureTimePerVolumeUnit mmHg_min_Per_mL = new PressureTimePerVolumeUnit("mmHg min/mL");
  public static PressureTimePerVolumeUnit mmHg_min_Per_L = new PressureTimePerVolumeUnit("mmHg min/L");
  public static PressureTimePerVolumeUnit Pa_s_Per_m3 = new PressureTimePerVolumeUnit("Pa s/m^3");

  public static PressureTimePerVolumeUnit FromString(string u)
  {
    if (u == PressureTimePerVolumeUnit.cmH2O_s_Per_L.ToString())
      return PressureTimePerVolumeUnit.cmH2O_s_Per_L;
    if (u == PressureTimePerVolumeUnit.cmH2O_s_Per_mL.ToString())
      return PressureTimePerVolumeUnit.cmH2O_s_Per_mL;
    if (u == PressureTimePerVolumeUnit.mmHg_s_Per_mL.ToString())
      return PressureTimePerVolumeUnit.mmHg_s_Per_mL;
    if (u == PressureTimePerVolumeUnit.mmHg_min_Per_mL.ToString())
      return PressureTimePerVolumeUnit.mmHg_min_Per_mL;
    if (u == PressureTimePerVolumeUnit.mmHg_min_Per_L.ToString())
      return PressureTimePerVolumeUnit.mmHg_min_Per_L;
    if (u == PressureTimePerVolumeUnit.Pa_s_Per_m3.ToString())
      return PressureTimePerVolumeUnit.Pa_s_Per_m3;

    throw new System.ArgumentException(u + " is not a know pressure time per volume unit");
  }
}

public class SEScalarPressureTimePerVolume : SEScalar
{
  protected PressureTimePerVolumeUnit unit;

  public SEScalarPressureTimePerVolume() : base()
  {

  }

  public SEScalarPressureTimePerVolume(SEScalarPressureTimePerVolume from)
  {
    Set(from);
  }

  public SEScalarPressureTimePerVolume(double value, PressureTimePerVolumeUnit unit)
  {
    SetValue(value, unit);
  }

  public void Set(SEScalarPressureTimePerVolume from)
  {
    if (from == null)
    {
      Invalidate();
      return;
    }
    SetValue(from.value, from.unit);
  }

  public new void SetValue(double value)
  {
    throw new System.NotSupportedException("You must provide a unit");
  }

  public void SetValue(double value, PressureTimePerVolumeUnit unit)
  {
    this.value = value;
    this.unit = unit;
  }

  public double GetValue(PressureTimePerVolumeUnit unit)
  {
    throw new System.NotImplementedException();
  }

  public PressureTimePerVolumeUnit GetUnit()
  {
    return unit;
  }

  public new string ToString()
  {
    return this.value + "(" + this.unit.ToString() + ")";
  }
}

File: src/cpp/cdm/properties/SEScalarPressureTimePerVolume.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarPressureTimePerVolume.h"

const PressureTimePerVolumeUnit PressureTimePerVolumeUnit::cmH2O_s_Per_L("cmH2O s/L");
const PressureTimePerVolumeUnit PressureTimePerVolumeUnit::cmH2O_s_Per_mL("cmH2O s/mL");
const PressureTimePerVolumeUnit PressureTimePerVolumeUnit::mmHg_s_Per_mL("mmHg s/mL");
const PressureTimePerVolumeUnit PressureTimePerVolumeUnit::mmHg_min_Per_mL("mmHg min/mL");
const PressureTimePerVolumeUnit PressureTimePerVolumeUnit::mmHg_min_Per_L("mmHg min/L");
const PressureTimePerVolumeUnit PressureTimePerVolumeUnit::Pa_s_Per_m3("Pa s/m^3");

bool PressureTimePerVolumeUnit::IsValidUnit(const std::string& unit)
{
  if (cmH2O_s_Per_L.GetString().compare(unit) == 0)
    return true;
  if (cmH2O_s_Per_mL.GetString().compare(unit) == 0)
    return true;
  if (mmHg_s_Per_mL.GetString().compare(unit) == 0)
    return true;
  if (mmHg_min_Per_mL.GetString().compare(unit) == 0)
    return true;
  if (mmHg_min_Per_L.GetString().compare(unit) == 0)
    return true;
  if (Pa_s_Per_m3.GetString().compare(unit) == 0)
    return true;
  return false;
}

const PressureTimePerVolumeUnit& PressureTimePerVolumeUnit::GetCompoundUnit(const std::string& unit)
{
  if (cmH2O_s_Per_L.GetString().compare(unit) == 0)
    return cmH2O_s_Per_L;
  if (cmH2O_s_Per_mL.GetString().compare(unit) == 0)
    return cmH2O_s_Per_mL;
  if (mmHg_s_Per_mL.GetString().compare(unit) == 0)
    return mmHg_s_Per_mL;
  if (mmHg_min_Per_mL.GetString().compare(unit) == 0)
    return mmHg_min_Per_mL;
  if (mmHg_min_Per_L.GetString().compare(unit) == 0)
    return mmHg_min_Per_L;
  if (Pa_s_Per_m3.GetString().compare(unit) == 0)
    return Pa_s_Per_m3;
  throw CommonDataModelException(unit + " is not a valid PressureTimePerVolume unit");
}

template class SEScalarQuantity<PressureTimePerVolumeUnit>;
